internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -162,6 +163,19 @@ func (av agentVersion) String() string {
 func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string, action *fleetapi.ActionUpgrade, det *details.Details, skipVerifyOverride bool, skipDefaultPgp bool, pgpBytes ...string) (_ reexec.ShutdownCallbackFn, err error) {
 	u.log.Infow("Upgrading agent", "version", version, "source_uri", sourceURI)
 
+	currentVersion := agentVersion{
+		version:  release.Version(),
+		snapshot: release.Snapshot(),
+		hash:     release.Commit(),
+	}
+
+	// check version before download
+	same, _ := isSameVersion(u.log, currentVersion, packageMetadata{}, version)
```

Comment:
Yes, but it should work for non-snapshot attempts, right?

Reply:
@pchila  so in this case this will fail because hash is unpopulated right? but then later it is extracted and checked again on line 224. so we lose a bit of time but will not perform upgrade

Reply:
Here we are relying on the fact that the current version has a not empty hash compared to the zero value in `packageMetadata{}`  so I don't see how it can ever detect this to be the same version if the has in packageMetadata is ""

I guess we will never get true from here... Did you test this ?

Reply:
you're correct, I forgot that comparing currentVersion vs version would also compare the hashes

Reply:
I've added a new func that does this comparison correctly.